### DISCUSSION TITLE
Suggestion: Add starting/stopping/playback to the Overview WIKI

### DISCUSSION THREAD

**Discussion by codegrue** [pschatzmann/arduino-audio-tools#1337] [Ideas]
I've been playing with this for a few hours and it's not obvious to me how to play another source. Actually,  it's not even obvious what starts playback. Is it the copier object being created? Is it out.begin()? What if I want to play a different sound after the first completes? Or end the first mid-stream and switch to another?

I did look through *most* of the examples and did not see one that was switching tracks. Even the player example didn't seem to show this.

Thanks!

**Comment by pschatzmann** [comment]
Thanks you for your feed-back.

I think you missed to read or understand the basic fundamentals that are explained in the first chapter of the Wiki!

For your question how to switch the source, I can't really give any answer because it depends on the case and usually you have many different alternatives to handle this. E.g. if you have multiple sources you could set up a copier for each and switch the copier, or you can just reconfigure the copier and define a new source and destination.

The AudioKit has some defined buttons, so have a look at the AudioPlayer examples: e.g. [this](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/player-sd-audiokit/player-sd-audiokit.ino). 

The AudioActions class can be used for handling button actions. But in the end this is an Audio library and not a button handling library...

**Comment by codegrue** [comment]
I reviewed that but it's not exactly what I am looking for. I don't actually want to make a music player, rather I am looking to play sounds based on other events (picture sound effects in a video game). So I don't want "previous" or "next" like the AudioPlayer delivers, but instead just play a random sound from PROGMEM preferrably.

I have tried the following when events trigger:

```
switch (effectToPlay)
{
case 1:
  copier.begin(out, sound1);
case 2:
  copier.begin(out, sound2);
case 3
  copier.begin(out, sound3);
}
```

but once they play through, the sounds won't trigger again. It's like they are at the end. I want to reset the copier and play whatever arbitrary sound is needed at that time

**Reply by pschatzmann** [reply]
As explained in the Wiki you can check how many bytes the copy() was processing: If you get 0 you are at the end and you can call begin on the source again to restart or better activate the looping on the MemoryStream in the setup so this is automatically done for you.

**Reply by codegrue** [reply]
Thank you for the reply. You got me on the right track. I solved it with this:

```
switch (effectToPlay)
{
case 1:
    sound1.begin();
    copier.begin(out, sound1);
case 2:
    sound2.begin();
    copier.begin(out, sound2);
case 3
    sound3.begin();
    copier.begin(out, sound3);
}
```

I just feel a section in the wiki overview for "managing playback" would be helpful. But maybe I just wasn't getting the concepts you have desribed there under Arduino Audio Streams. Thank you for the help and this is a great library!

**Reply by pschatzmann** [reply]
yes, this will always play from the start of the Stream...

**Reply by codegrue** [reply]
I found that if this technique  is used to play another sound while the previous one has not finished, there is a "pop" before the switch. I tried adding a copier.end(); before the new begin but that didn't fix it. Is there a way to "stop playing the current sound gracefully?"

**Reply by pschatzmann** [reply]
The pop happens when the audio signal is near it's peak and you suddenly stop or start it.

You could implement your own fade in and fade out functionality based on managing the volume yourself or you can use the existing FadeConverter which implements a fade in and fade out. There is also a PoppingSoundRemover which tries to remove the sound after the last zero crossing.

The copier supports a converter as parameter - tough I have never tried this for a fade in or fade out...

**Comment by tomliekens** [comment]
dear codegrue, do you mind sharing the full code where you use this switch (effectToPlay)... snippet in pls? i would like to change the mp3 file based on received uart strings like a dfplayer mini does. but i don't yet see how to do this in this great library.
Maybe i can build further on your code...
many many thx

**Reply by codegrue** [reply]
Here is my complete sound component:

```
#include "sound.h"

#define SOUND_AIRPLAME_START "/sounds/airplane-start.wav"
#define SOUND_AIRPLANE_ALARM "/sounds/airplane-alarm.wav"
#define SOUND_AIRPLANE_PA "/sounds/airplane-pa-short.wav"
#define SOUND_F_IFR "/sounds/F-IFR.wav"
#define SOUND_F_LIFR "/sounds/F-LIFR.wav"
#define SOUND_F_MVFR "/sounds/F-MVFR.wav"
#define SOUND_F_NA "/sounds/F-NA.wav"
#define SOUND_F_PVFR "/sounds/F-PVFR.wav"
#define SOUND_F_VFR "/sounds/F-VFR.wav"
#define SOUND_F_ROGER "/sounds/F-Roger.wav"
#define SOUND_M_IFR "/sounds/M-IFR.wav"
#define SOUND_M_LIFR "/sounds/M-LIFR.wav"
#define SOUND_M_MVFR "/sounds/M-MVFR.wav"
#define SOUND_M_NA "/sounds/M-NA.wav"
#define SOUND_M_PVFR "/sounds/M-PVFR.wav"
#define SOUND_M_VFR "/sounds/M-VFR.wav"
#define SOUND_M_ROGER "/sounds/M-Roger.wav"

AudioInfo audioInfo(SAMPLE_RATE, AUDIO_CHANNELS, BITS_PER_SAMPLE);
File wavFile;

// data flow: wavFile -> decoder -> encoder -> volume -> fader -> i2s
I2SStream i2s; // final output of stream
FadeStream fader(i2s);
VolumeStream volume(fader); // intermediate to set the volume
WAVDecoder decoder;
EncodedAudioOutput encoder(&volume, &decoder); // Decoding stream
StreamCopy copier;

bool g_repeat = false;

void Sound::setup()
{
    if (BOARD_SUPPORTS_SPEAKER)
    {
        ESP_LOGI(MODULE_NAME_APP_SETUP, "Sound - Starting");

        if (DEBUG_SHOW_SOUND_INFO)
            AudioLogger::instance().begin(Serial, AudioLogger::Info);
        else
            AudioLogger::instance().begin(Serial, AudioLogger::Warning);

        I2SConfig config = i2s.defaultConfig(TX_MODE);
        config.copyFrom(audioInfo);
        config.pin_bck = I2S_BCLK;
        config.pin_data = I2S_DOUT;
        config.pin_ws = I2S_LRC;
        i2s.begin(config);

        fader.setAudioInfo(audioInfo);

        volume.begin(config); // we need to provide the bits_per_sample and channels

        float startingVolume = g_settings.getVoiceVolume() / 100.0;
        volume.setVolume(startingVolume);

        encoder.begin(); // reset sound at beginning of data

        copier.begin(encoder, wavFile); // activate copier
        copier.setMinCopySize(1);       // allow partial frames
    }

    SerialHelpers::logHeapInfo("after sound setup");
}

void Sound::stop()
{
    if (BOARD_SUPPORTS_SPEAKER)
    {
        g_repeat = false;
        wavFile.close();

        SerialHelpers::logHeapInfo("after sound stop");
    }
}

void Sound::play(EmbeddedSounds soundName, bool alarmVolume, bool repeat)
{
    if (BOARD_SUPPORTS_SPEAKER)
    {
        // ESP_LOGI(MODULE_NAME_SOUND, "Playing Sound");

        bool useMaleVoice = g_settings.getVoiceMale();

        const char *filePath = SOUND_AIRPLANE_PA;

        switch (soundName)
        {
        case AIRPLANE_PA:
            filePath = SOUND_AIRPLANE_PA;
            break;
        case AIRPLANE_START:
            filePath = SOUND_AIRPLAME_START;
            break;
        case AIRPLANE_ALARM:
            filePath = SOUND_AIRPLANE_ALARM;
            break;
        case ROGER:
            filePath = (useMaleVoice) ? SOUND_M_ROGER : SOUND_F_ROGER;
            break;
        case IFR:
            filePath = (useMaleVoice) ? SOUND_M_IFR : SOUND_F_IFR;
            break;
        case VFR:
            filePath = (useMaleVoice) ? SOUND_M_VFR : SOUND_F_VFR;
            break;
        case PVFR:
            filePath = (useMaleVoice) ? SOUND_M_PVFR : SOUND_F_PVFR;
            break;
        case MVFR:
            filePath = (useMaleVoice) ? SOUND_M_MVFR : SOUND_F_MVFR;
            break;
        case LIFR:
            filePath = (useMaleVoice) ? SOUND_M_LIFR : SOUND_F_LIFR;
            break;
        case NA:
            filePath = (useMaleVoice) ? SOUND_M_NA : SOUND_F_NA;
            break;
        }

        g_repeat = repeat;

        wavFile = Files::openFileOnSPIFFS(filePath);

        float currentVolume = ((alarmVolume) ? g_settings.getAlarmVolume() : g_settings.getVoiceVolume()) / 100.0;
        volume.setVolume(currentVolume); // set the playback volume

        SerialHelpers::logHeapInfo("before playing new sound");
    }
}

bool Sound::isRunning()
{
    if (BOARD_SUPPORTS_SPEAKER)
    {
        return copier.isActive();
    }

    return false;
}

void Sound::loop()
{
    if (BOARD_SUPPORTS_SPEAKER)
    {
        try
        {
            if (wavFile.available()   // file is active
                && wavFile.size() > 0 // file has data
            )
            {
                int size = copier.copy();

                // there is a bug where sounds sometimes don't complete,
                // this code is a workaround to force completion
                if (size == 0)
                {
                    ESP_LOGW(MODULE_NAME_SOUND, "copier.copy() %d %d %d %d", size, wavFile.available(), wavFile.position(), wavFile.size());

                    wavFile.close();
                }
            }
            else if (g_repeat)
            {
                wavFile.seek(0); // set sound to play again
            }
        }
        catch (const std::exception &e)
        {
            // don't allow an exception here to crash the system
            ESP_LOGE(MODULE_NAME_SOUND, "%s", e.what());
        }
    }
}
```

**Reply by CptDoucheWizard** [reply]
I have a similar application and have been exploring switching tracks and all has been going well.
What I can't work out is removing the gap between tracks when switching. I have tried a method based on codegru's approach where the file is opened each time the track switches (which intuitively sounds slow), and I have also tried loading in all the tracks, and then switching which is played by feeding each track seperately into a copier.begin statement (which sounds a bit RAM intensive).

With both approaches I get a noticable delay between the end of one track and the beginning of another. The only other thing I can think of is to have all the different sounds on the one track, and then seek to different positions for different sounds. I suspect that will have almost no delay at all, but is a messier implementation in my mind.

Any thoughts on the 'correct' workflow for minimising delay between tracks?
